From 6f283e3642856b4a9cd1a9ed7dc637ad552f5869 Mon Sep 17 00:00:00 2001
From: Gleb Mazovetskiy <glex.spb@gmail.com>
Date: Wed, 13 Jan 2021 19:20:42 +0000
Subject: [PATCH 2/4] cmake: OSG_STATIC: Fix plugin linking

OSG libraries that are used by plugins must also be linked with
`--whole-archive` to avoid undefined references at link time.
---
 apps/openmw/CMakeLists.txt | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/apps/openmw/CMakeLists.txt b/apps/openmw/CMakeLists.txt
index d35d7c4a3..65061f6d5 100644
--- a/apps/openmw/CMakeLists.txt
+++ b/apps/openmw/CMakeLists.txt
@@ -121,8 +121,14 @@ include_directories(
     ${FFmpeg_INCLUDE_DIRS}
 )
 
-target_link_libraries(openmw
+# These OSG libraries are used by OSG plugins.
+# They have to be linked with --whole-archive in a static build.
+set(_osg_libraries_used_by_osg_plugins
     ${OSG_LIBRARIES}
+    ${OSGTEXT_LIBRARIES}
+)
+
+target_link_libraries(openmw
     ${OSGPARTICLE_LIBRARIES}
     ${OSGUTIL_LIBRARIES}
     ${OSGDB_LIBRARIES}
@@ -144,12 +150,15 @@ target_link_libraries(openmw
 )
 
 if(OSG_STATIC)
-    set(_osg_plugins -Wl,--whole-archive)
+    # We use `--whole-archive` because plugins use registration.
+    set(_osg_plugins -Wl,--whole-archive ${_osg_libraries_used_by_osg_plugins})
     foreach(PLUGIN_NAME ${USED_OSG_PLUGINS})
         set(_osg_plugins ${_osg_plugins} ${OSGPlugins_LIB_DIR}/lib${PLUGIN_NAME}.a)
     endforeach()
     set(_osg_plugins ${_osg_plugins} -Wl,--no-whole-archive)
     target_link_libraries(openmw ${_osg_plugins})
+else()
+    target_link_libraries(openmw ${_osg_libraries_used_by_osg_plugins})
 endif(OSG_STATIC)
 
 if (ANDROID)
-- 
2.27.0

